tool/tsh/exec_command_test.go

```diff
@@ -0,0 +1,221 @@
+				exec := newExecCommand(app)
+				exec.arguments = []string{"-c", "exit 0"}
+				exec.shell = "bash"
+				return exec
+			},
+			args:    args{cf: &CLIConf{}},
+			wantErr: false,
+		},
+		{
+			name: "exit code 1, error",
+			cmd: func(t *testing.T) *execCommand {
+				app := kingpin.New("my test app", "")
+				exec := newExecCommand(app)
+				exec.arguments = []string{"-c", "exit 1"}
+				exec.shell = "bash"
+				return exec
+			},
+			args:    args{cf: &CLIConf{}},
+			wantErr: true,
+		},
+		{
+			name: "check environment",
+			cmd: func(t *testing.T) *execCommand {
+				app := kingpin.New("my test app", "")
+				exec := newExecCommand(app)
+				exec.arguments = []string{"-c", `
+cd $0
+echo $TSH_COMMAND >> tsh_command.env
+echo $TSH >> tsh.env`, tmp}
+				exec.shell = "bash"
+				return exec
+			},
+			args: args{cf: &CLIConf{executablePath: "/bin/tsh"}},
+			postCheck: func(t *testing.T) {
+				checkFile := func(fn string, contents string) {
+					bytes, err := os.ReadFile(path.Join(tmp, fn))
+					require.NoError(t, err)
+					require.Equal(t, contents, string(bytes))
+				}
+
+				checkFile("tsh_command.env", "/bin/bash\n")
+				checkFile("tsh.env", "/bin/tsh\n")
+			},
+			wantErr: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			exec := tt.cmd(t)
+			err := exec.runCommand(tt.args.cf)
+
+			if tt.wantErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+
+			if tt.postCheck != nil {
+				tt.postCheck(t)
+			}
+		})
+	}
+}
+
+func Test_findShell(t *testing.T) {
+	mkshell := func(fn string) {
+		// use existing `sh` binary for tests.
+		shPath, err := os_exec.LookPath("sh")
+		require.NoError(t, err)
+
+		file, err := os.ReadFile(shPath)
+		require.NoError(t, err)
+
+		err = os.WriteFile(fn, file, 0777)
+		require.NoError(t, err)
+	}
+
+	tmp := t.TempDir()
+
+	tests := []struct {
+		name string
+		init func(t *testing.T) string
+		want string
+	}{
+		{
+			name: "return custom shell",
+			init: func(t *testing.T) string {
+				err := os.MkdirAll(path.Join(tmp, "custom"), 0777)
+				require.NoError(t, err)
+
+				dummyShell := path.Join(tmp, "custom", "dummysh")
+				mkshell(dummyShell)
+
+				return dummyShell
+			},
+			want: path.Join(tmp, "custom", "dummysh"),
+		},
+
+		{
+			name: "return $SHELL",
+			init: func(t *testing.T) string {
+				err := os.MkdirAll(path.Join(tmp, "SHELL"), 0777)
+				require.NoError(t, err)
+
+				dummyShell := path.Join(tmp, "SHELL", "dummysh")
+				mkshell(dummyShell)
+
+				t.Setenv("SHELL", dummyShell)
+
+				return ""
+			},
+			want: path.Join(tmp, "SHELL", "dummysh"),
+		},
+
+		{
+			name: "return sh from PATH",
+			init: func(t *testing.T) string {
+				err := os.MkdirAll(path.Join(tmp, "PATH_SH"), 0777)
+				require.NoError(t, err)
+
+				dummyShell := path.Join(tmp, "PATH_SH", "sh")
+				mkshell(dummyShell)
+
+				// unset shell, set path
+				t.Setenv("SHELL", "")
+				t.Setenv("PATH", path.Join(tmp, "PATH_SH"))
```

Comment:
Yep, the cleanup is automatic. In case of `t.SetEnv` the scheduling is also important: the env vars are essentially a global variables for the given process, so we shouldn't be setting those without making sure there is only a single test running or that it is somehow isolated. Anyway, these framework-given helpers are not that much different from the regular functions in standard lib, except for the cleanup bit.